### ISSUE TITLE
Prediction score extremely low, despite high (~94%) validation accuracy with bottleneck features (VGG16)

### ISSUE THREAD

**Issue by KeironO** [keras-team/keras#8114]
Hi there,

I'm trying to utilise "bottleneck features" produced from VGG16 for a new classification task with limited data.

Right now, I have the following (please, excuse code quality):

```python
import numpy as np
from keras.preprocessing.image import ImageDataGenerator
from keras.applications import VGG16
from PIL import Image
from keras.preprocessing import image
from keras.utils import to_categorical
from keras.models import Sequential, Model
from keras.layers import Dropout, Flatten, Dense, Input
from sklearn.metrics import classification_report
import os


height = 150
width = 150
batch_size = 16

train_dir = "/home/keo7/Data/PVS/train/"
val_dir = "/home/keo7/Data/PVS/val/"
test_dir = "/home/keo7/Data/PVS/test/"

nb_train_samples = 6000
nb_validation_samples = 1000


def get_bottleneck_features():
    data_generator = ImageDataGenerator(rescale=1. / 255)

    vgg_16 = VGG16(include_top=False, weights="imagenet")

    generator = data_generator.flow_from_directory(
        train_dir,
        target_size=[width, height],
        batch_size=batch_size,
        class_mode=None,
        shuffle=False
    )

    train_features = vgg_16.predict_generator(generator,
                                              nb_train_samples // batch_size,
                                              verbose=1)
    np.save("features_train.npy", train_features)

    generator = data_generator.flow_from_directory(val_dir,
                                                   target_size=[width, height],
                                                   batch_size=batch_size,
                                                   class_mode=None,
                                                   shuffle=False)

    val_features = vgg_16.predict_generator(generator, nb_validation_samples // batch_size)
    np.save("features_val.npy", val_features)

    return generator.class_indices


def train_top_model():
    train_data = np.load("features_train.npy")
    train_labels = np.array([[0] * 600, [1] * 600, [2] * 600, [3] * 600, [4] * 600,
                             [5] * 600, [6] * 600, [7] * 600, [8] * 600, [9] * 600])
    train_labels_one_hot = to_categorical(train_labels, 10)

    validation_data = np.load("features_val.npy")
    validation_labels = np.array([[0] * 100, [1] * 100, [2] * 100, [3] * 100, [4] * 100,
                                  [5] * 100, [6] * 100, [7] * 100, [8] * 100, [9] * 100])
    validation_labels_one_hot = to_categorical(validation_labels, 10)

    model = Sequential()
    model.add(Flatten(input_shape=train_data.shape[1:]))
    model.add(Dense(256, activation="relu"))
    model.add(Dropout(0.5))
    model.add(Dense(len(train_labels), activation="softmax"))

    model.compile(optimizer='rmsprop', loss='categorical_crossentropy', metrics=['accuracy'])

    model.fit(train_data, train_labels_one_hot[:len(train_data)],
              epochs=5,
              batch_size=batch_size,
              validation_data=(validation_data, validation_labels_one_hot[:len(validation_data)])
              )

    model.save_weights("fc_model.h5")

def create_final_model():
    base_model = VGG16(include_top=False, weights='imagenet', input_shape=(150, 150, 3))
    top_model = Sequential()
    top_model.add(Flatten(input_shape=base_model.output_shape[1:]))
    top_model.add(Dense(256, activation='relu'))
    top_model.add(Dropout(0.5))
    top_model.add(Dense(10, activation='softmax'))
    top_model.load_weights("fc_model.h5")

    final_input = Input(shape=(150, 150, 3))
    x = base_model(final_input)
    result = top_model(x)
    return Model(input=final_input, output=result)

def predict_image_class(test_image, classes, final_model):
    img = Image.open(test_image)
    img = img.resize((height, width))
    x = image.img_to_array(img)
    x = np.expand_dims(x, axis=0)
    preds = final_model.predict(x / 255)[0]
    return classes[np.argmax(preds)]

if __name__ == "__main__":

    get_bottleneck_features()
    train_top_model()

    classes = []
    train_path = train_dir
    for item in os.listdir(train_path):
        if os.path.isdir(os.path.join(train_path, item)):
            if item not in classes:
                classes.append(item)


    y_true = []
    y_pred = []

    final_model = create_final_model()

    for d, sd, _ in os.walk(train_dir):
        for s in sd:
            for file in os.listdir(os.path.join(d, s)):
                prediction = predict_image_class(os.path.join(d,s,file), classes, final_model)
                y_true.append(s)
                y_pred.append(prediction)

    print classification_report(y_true, y_pred)
```

Which returns the following output:

```
/home/keo7/.virtualenvs/deeplearning/bin/python /home/keo7/Projects/AgriDoctor/classifier/test.py
Using Theano backend.
Using cuDNN version 6021 on context None
Mapped name None to device cuda: GeForce GTX 1060 (0000:01:00.0)
Found 6000 images belonging to 10 classes.
  1/375 [..............................] - ETA: 144s
...
375/375 [==============================] - 80s    
Found 1000 images belonging to 10 classes.
Train on 6000 samples, validate on 992 samples
Epoch 1/5
  16/6000 [..............................] - ETA: 0s - loss: 2.9712 - acc: 0.1250
...
6000/6000 [==============================] - 0s - loss: 0.9426 - acc: 0.7107 - val_loss: 0.3441 - val_acc: 0.8810
Epoch 2/5
  16/6000 [..............................] - ETA: 0s - loss: 0.6666 - acc: 0.7500
...
6000/6000 [==============================] - 0s - loss: 0.4603 - acc: 0.8427 - val_loss: 0.2814 - val_acc: 0.8931
Epoch 3/5
  16/6000 [..............................] - ETA: 0s - loss: 0.6545 - acc: 0.8750
...
6000/6000 [==============================] - 0s - loss: 0.3573 - acc: 0.8798 - val_loss: 0.1961 - val_acc: 0.9244
Epoch 4/5
  16/6000 [..............................] - ETA: 0s - loss: 0.1804 - acc: 0.9375
...
6000/6000 [==============================] - 0s - loss: 0.3017 - acc: 0.9062 - val_loss: 0.2160 - val_acc: 0.9214
Epoch 5/5
  16/6000 [..............................] - ETA: 0s - loss: 0.1237 - acc: 0.9375
 ...
6000/6000 [==============================] - 0s - loss: 0.2462 - acc: 0.9218 - val_loss: 0.1746 - val_acc: 0.9456
/home/keo7/Projects/AgriDoctor/classifier/test.py:97: UserWarning: Update your `Model` call to the Keras 2 API: `Model(outputs=Softmax.0, inputs=/input_3)`
  return Model(input=final_input, output=result)

             precision    recall  f1-score   support

       c_15       0.00      0.00      0.00       600
       c_16       0.00      0.00      0.00       600
       c_24       0.00      0.00      0.00       600
       c_25       0.00      0.00      0.00       600
       c_28       0.00      0.00      0.00       600
        c_3       0.00      0.00      0.00       600
       c_30       0.00      0.00      0.00       600
       c_32       0.00      0.00      0.00       600
       c_33       0.00      0.00      0.00       600
       c_35       0.00      0.00      0.00       600

avg / total       0.00      0.00      0.00      6000
```

I'm fairly accustom to Keras and deep learning in general, but I've yet to use this "bottleneck" technique. I believe the problem is likely to be related to the corresponding of predicted output to informative labels, but I'm struggling to find a solution.

Any help would be fantastic!

Thanks,

Keiron.
